### PR TITLE
Add new UI client

### PR DESCRIPTION
**Pull request by jlesquembre**
Built on top of [apiUpdates branch](https://github.com/tweag/nixpkgs-graph-explorer/tree/dorranh/apiUpdates). This PR is only about the JS client, I'm not including any of the changes from `apiUpdates` branch

I think it should work with current `main`, but I didn't test it (We want to merge `apiUpdates` anyways, right?)

### CODE REVIEW

**Comment by GuillaumeDesforges** [comment]
Could you update the README to help build/use the new UI?

File: js-client/README.md
Comment (GuillaumeDesforges) [inline js-client/README.md:1]: :heart:

File: js-client/src/nix-search.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  constructor() {
+    super();
+    this.worker.addEventListener("message", (msg: MessageEvent<Operation>) => {
+      const { type, data } = msg.data;
+      if (type === "search") {
+        this.pkgs = data;
+        this.loading = false;
+      }
+    });
+    this.worker.postMessage({ type: "search", data: "" });
```
Comment (GuillaumeDesforges) [inline js-client/src/nix-search.ts:40]: Why is this needed?

File: js-client/src/nix-search.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+  @state()
+  pkgs: string[] = [];
+
+  @state()
+  loading = true;
+
+  constructor() {
+    super();
+    this.worker.addEventListener("message", (msg: MessageEvent<Operation>) => {
```
Comment (GuillaumeDesforges) [inline js-client/src/nix-search.ts:33]: What's the garantee that we have `msg: MessageEvent<Operation>`?

https://github.com/tweag/nixpkgs-graph-explorer/pull/37/files#diff-8a6044e9823d1a782d9032e7b7d939d39de15bf0278f0a1d723492d43cad04aeR92

does not seem typed at all

File: js-client/src/packages.ts
Comment (GuillaumeDesforges) [inline js-client/src/packages.ts:1]: I won't review it since it's temporary

**Review by GuillaumeDesforges — COMMENTED** [review]
Nice!

This is a good PR to scaffold the new frontend SPA.

Did you use a formatter? If not I suggest we use Prettier.

File: js-client/src/nix-search.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import "@shoelace-style/shoelace/dist/components/spinner/spinner.js";
+import "@shoelace-style/shoelace/dist/components/alert/alert.js";
+import "@shoelace-style/shoelace/dist/components/icon/icon.js";
+
+interface Operation {
+  type: "search" | "graph";
+  data: string[];
+}
+
+type EventInput = Event & { target: HTMLInputElement };
```
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/src/nix-search.ts:17]: Nope :( 
https://github.com/microsoft/TypeScript/issues/299

File: js-client/package-lock.json
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/package-lock.json:1]: In my experience, yarn works better with `...2nix` tools. Since we don't use it any 2nix tools, npm works fine. If you see a reason to move to yarn, I don't have anything against it.

File: js-client/src/app-main.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import { LitElement, css, html } from "lit";
+import { customElement, query } from "lit/decorators.js";
+import "@shoelace-style/shoelace/dist/themes/light.css";
+import "./nix-search.ts";
+
+import cytoscape from "cytoscape";
+import dagre from "cytoscape-dagre";
+
+const API_URL = "/api";
```
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/src/app-main.ts:9]: Sure, now take the value from the env variable `API_URL`, defaults to `/api` if not defined

File: js-client/src/app-main.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+interface GraphEv {
```
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/src/app-main.ts:11]: It is the custom event payload, moved it to `nix-search` file and renames to `ClickItemPayload`

File: js-client/src/app-main.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      query: `g.V().filter{it.get().value('pname').matches('${pkgName}')}.repeat(outE().otherV().simplePath()).until(__.not(outE().simplePath())).path().by('pname').by(label)`,
+    }),
+  });
+  const r = await response.json();
+  console.log(r);
+  const data = r.cyto["graph-data"];
+  cytoscape.use(dagre);
```
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/src/app-main.ts:29]: Good catch!

File: js-client/package-lock.json
[the remark this answers is not in the document]
Comment (GuillaumeDesforges) [reply js-client/package-lock.json:1]: Historically, npm didn't really have a way to lock dependencies for real.
If `js-client/package-lock.json` is really a good lockfile, no need to switch.

File: js-client/src/app-main.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      <nix-search @graph=${this.updateGraph}> </nix-search>
+      <div id="cy-container">
+        <div id="cy"></div>
+      </div>
+    `;
+  }
+
+  private updateGraph(ev: CustomEvent<GraphEv>) {
+    getGraph(ev.detail.name, this._cy);
+  }
```
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/src/app-main.ts:59]: `nix-search` component emits a custom event when the user clicks on a result. We want to catch the event to render a new graph.

File: js-client/src/nix-search.ts
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/src/nix-search.ts:1]: I think that was the idea of the Lit team. But the internals are very different

File: js-client/src/nix-search.ts (hunk 1 above)
Comment (jlesquembre) [reply js-client/src/nix-search.ts:40]: It's used to make an initial query and display a list of packages. Not really needed, but then the user doesn't see any packages initially

File: js-client/src/nix-search.ts (hunk 2 above)
Comment (jlesquembre) [reply js-client/src/nix-search.ts:33]: I picked poor names. Those events are not related, I'm changing the name of the custom event to avoid confusions.

File: js-client/package-lock.json
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/package-lock.json:1]: I remember some corner cases with npm lock format, but that was a long time ago, maybe those are solved already.  It we hit any issue, we can always switch, it's a relatively easy change.

**Comment by jlesquembre** [comment]
> Did you use a formatter? If not I suggest we use Prettier.

I use prettier. Once we have a `flake.nix` we could add a command to check the format.

File: js-client/src/nix-search.ts (hunk 1 above)
Comment (GuillaumeDesforges) [reply js-client/src/nix-search.ts:49]: Right, thanks. I would add a comment then
```suggestion
    # trigger a request to get an initial list of packages displayed
    this.worker.postMessage({ type: "search", data: "" });
```

File: js-client/src/app-main.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  private async updateGraph(ev: CustomEvent<ClickItemPayload>) {
+    try {
+      this._error = false; // Remove old errors
+      const cy = await this._cy;
+      const graphData = await getGraph(ev.detail.name);
+      renderCyGraph(graphData, cy);
+    } catch {
+      this._error = true;
+    }
+  }
```
Comment (GuillaumeDesforges) [inline js-client/src/app-main.ts:69]: It seems confusing to me to have a method `updateGraph` which is an event handler which is passed to `nix-search`, instead of being bound to the `AppMain` it belongs.

Would it be possible

to have a `interface UpdateGraphRequestPayload { name: string; }` type in `app-main.ts`

then to have an event handler in `AppMain`

```suggestion
  private async handleUpdateGraphRequest(ev: CustomEvent<UpdateGraphRequestPayload>) {
      // will be called when a request to update the graph view is dispatched
```

then to have the logic to listen to such events stay in `AppMain` and not leak to other components?

Then other components like `NixSearch` can dispatch events without being tightly coupled.

File: js-client/src/app-main.ts [hunk 9]
```diff
@@ -1,18 +1,16 @@
...
 import dagre from "cytoscape-dagre";
+import cytoscape from "cytoscape";
 
-const API_URL = "/api";
-
-interface GraphEv {
-  name: string;
-}
+const API_URL = __API_URL__;
+cytoscape.use(dagre);
```
Comment (GuillaumeDesforges) [inline js-client/src/app-main.ts:11]: Can that fail? If so, can leaving that in the global scope lead to unrendered components?

**Comment by dorranh** [comment]
FYI - @jlesquembre, #39 has now been merged. Also, one question about these changes. From what I can tell, these only include the table and graph views and not the query editor which we have in the prototype deployed to https://nixpkgs-graph.app/. Would it be possible to somehow include that as well (e.g. as a tab)? I think having the table view as the default as you have it here is nice though.

**Comment by GuillaumeDesforges** [comment]
@dorranh I suggest we don't do that in this PR, get it merged ASAP so we can start on all the UI features we want in smaller, short-lived  feature branch. WDYT?

File: js-client/src/nix-search.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+import "@shoelace-style/shoelace/dist/components/button/button.js";
+import "@shoelace-style/shoelace/dist/components/menu/menu.js";
+import "@shoelace-style/shoelace/dist/components/menu-item/menu-item.js";
+import "@shoelace-style/shoelace/dist/components/input/input.js";
+import "@shoelace-style/shoelace/dist/components/spinner/spinner.js";
+import "@shoelace-style/shoelace/dist/components/alert/alert.js";
+import "@shoelace-style/shoelace/dist/components/icon/icon.js";
+
+interface Operation {
+  type: "search" | "graph";
```
[the remark this answers is not in the document]
Comment (jlesquembre) [reply js-client/src/nix-search.ts:18]: I think so, at least I don't see how to use different TS types here. 
The only way to interact with a worker is with [`postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage). First argument, `message`, can be anything. I'm defining my own type here

File: js-client/src/app-main.ts (hunk 8 above)
Comment (jlesquembre) [reply js-client/src/app-main.ts:69]: `app-main` is not passing anything to `nix-search`. `nix-search` is not coupled to `app-main`, it doesn't know anything about the parent component. When the user click on one of the packages, it emits a custom event with the package's name. Parent elements can decide if listen or ignore the event. 

Maybe there is some confusion because I called two different functions `updateGraph` in `app-main` and in `nix-search`: 
https://github.com/tweag/nixpkgs-graph-explorer/blob/9847650718357b20bb28987a7b6cc8ca5ff8f70b/js-client/src/nix-search.ts#L94
Those functions are independent, and thinking again about it, I should rename the function in `nix-search`

Once the logic is more complex, we should probably move the update graph logic to its own component.

File: js-client/src/app-main.ts (hunk 9 above)
Comment (jlesquembre) [reply js-client/src/app-main.ts:11]: You mean the `__API_URL__`? It would be always replaced by a string, with a fallback to `"/api"`:
https://github.com/tweag/nixpkgs-graph-explorer/blob/9847650718357b20bb28987a7b6cc8ca5ff8f70b/js-client/vite.config.ts#L26-L27

**Comment by jlesquembre** [comment]
@dorranh could you provide an example to call the `/gremlin` endpoint? With the latest API version I always get an empty response. Looks like something is wrong with my query but I don't find the reason.

**Comment by dorranh** [comment]
Hey ya you should be able to any valid Gremlin query, for example here is a request JSON body that works for getting `raw` data (but not the cytoscape graph input)

```json
{
    "query": "g.V()"
}
```

```json
{
    "raw": "[]"
}
```

For getting the cytoscape input you need to submit a query like the default one on nixpkgs-graph.app which returns a path, e.g.

```json
{
   "query": "g.V().filter{it.get().value('pname').matches('auto-multiple-choice')}.repeat(outE().otherV().simplePath()).times(2).path().by('pname').by(label)"
}
```

where `auto-multiple-choice` is the package name.

That should return an additional attribute called `cyto_data` in the response.

**Comment by jlesquembre** [comment]
> @jlesquembre could you rebase and make sure the UI you implemented works?

Done, I had to do some changes, now works with the new API

File: js-client/src/app-main.ts (hunk 9 above)
Comment (GuillaumeDesforges) [reply js-client/src/app-main.ts:11]: I meant

```
cytoscape.use(dagre);
```

in React I'd use some sort of context component (see [`useContext`](https://react.dev/reference/react/useContext)) with a [`useEffect`](https://react.dev/reference/react/useEffect) to call this `.use` only when relevant
